pkg/crypto/certmanager/manager.go

```diff
@@ -0,0 +1,451 @@
+	} else if err := cert.parse(); err != nil {
+		return nil, err
+	} else {
+		return cert, nil
+	}
+}
+
+func makeCert(chainDER [][]byte, key crypto.Signer) (*Certificate, error) {
+	if len(chainDER) < 1 {
+		return nil, errors.New("empty chain")
+	}
+	leaf, err := x509.ParseCertificate(chainDER[0])
+	if err != nil {
+		return nil, err
+	}
+	certPemBlock := &bytes.Buffer{}
+	for index, certDER := range chainDER {
+		if index > 0 {
+			fmt.Fprintln(certPemBlock)
+		}
+		err := pem.Encode(certPemBlock, &pem.Block{
+			Type:  "CERTIFICATE",
+			Bytes: certDER,
+		})
+		if err != nil {
+			return nil, err
+		}
+	}
+	keyPemBlock := &bytes.Buffer{}
+	switch key := key.(type) {
+	case *ecdsa.PrivateKey:
+		keyDER, err := x509.MarshalECPrivateKey(key)
+		if err != nil {
+			return nil, err
+		}
+		err = pem.Encode(keyPemBlock, &pem.Block{
+			Type:  "EC PRIVATE KEY",
+			Bytes: keyDER,
+		})
+		if err != nil {
+			return nil, err
+		}
+	default:
+		return nil, errors.New("unsupported private key type")
+	}
+	cert := &Certificate{
+		CertPemBlock: certPemBlock.Bytes(),
+		KeyPemBlock:  keyPemBlock.Bytes(),
+		tlsCert: tls.Certificate{
+			Certificate: chainDER,
+			PrivateKey:  key,
+			Leaf:        leaf,
+		},
+		notAfter:  leaf.NotAfter,
+		notBefore: leaf.NotBefore,
+	}
+	return cert, nil
+}
+
+func jitteryHour() time.Duration {
+	randByte := make([]byte, 1)
+	rand.Read(randByte)
+	return time.Hour + time.Second*time.Duration(randByte[0])
+}
+
+func (cert *Certificate) parse() error {
+	var err error
+	cert.tlsCert, err = tls.X509KeyPair(cert.CertPemBlock, cert.KeyPemBlock)
+	if err != nil {
+		return err
+	}
+	x509Cert, err := x509.ParseCertificate(cert.tlsCert.Certificate[0])
+	if err != nil {
+		return err
+	}
+	cert.notAfter = x509Cert.NotAfter
+	cert.notBefore = x509Cert.NotBefore
+	return nil
+}
+
+func (cert *Certificate) timeUntilRenewal(renewBefore float64) time.Duration {
+	if cert == nil {
+		return -1
+	}
+	lifetime := cert.notAfter.Sub(cert.notBefore)
+	if lifetime < time.Hour {
+		lifetime = jitteryHour()
+	}
+	return time.Until(cert.notAfter.Add(
+		-time.Duration(lifetime.Seconds()*renewBefore) * time.Second))
+}
+
+func newManager(names []string, certFilename, keyFilename string, locker Locker,
+	challengeType string, responder Responder, storer Storer,
+	renewBefore float64, caDirectoryURL string,
+	logger log.DebugLogger) (*CertificateManager, error) {
+	if challengeType == "" {
+		cert, err := loadCertificate(certFilename, keyFilename, logger)
+		if err != nil {
+			return nil, err
+		}
+		return &CertificateManager{certificate: cert}, nil
+	}
+	if locker == nil {
+		locker = nullLocker{}
+	}
+	if _, ok := supportedChallengeTypes[challengeType]; !ok {
+		return nil,
+			fmt.Errorf("challenge type: %s not supported", challengeType)
+	}
+	if renewBefore <= 0.0 {
+		randByte := make([]byte, 1)
+		if _, err := rand.Read(randByte); err != nil {
+			return nil, err
+		}
+		renewBefore = 0.32 + 0.02*float64(randByte[0])/256.0
```

Comment:
explain what are you trying to do here and why so many magic values.

Reply:
Done.